python/ray/data/impl/plan.py

```diff
@@ -348,3 +436,95 @@ def __call__(
         )
         assert isinstance(blocks, BlockList), blocks
         return blocks, stage_info
+
+
+def _rewrite_read_stages(
+    blocks: BlockList,
+    stats: DatasetStats,
+    stages: List[Stage],
+    dataset_uuid: str,
+) -> Tuple[BlockList, DatasetStats, List[Stage]]:
+    """Rewrites read stages into one-to-one stages, if needed."""
+    if _is_lazy(blocks) and stages:
+        blocks, stats, stage = _rewrite_read_stage(blocks)
+        stats.dataset_uuid = dataset_uuid
+        stages.insert(0, stage)
+    return blocks, stats, stages
+
+
+def _rewrite_read_stage(
+    in_blocks: LazyBlockList,
+) -> Tuple[BlockList, DatasetStats, Stage]:
+    """Rewrite the read stage to a OneToOne stage over read tasks as input.
+
+    For example, suppose the plan was [Read -> MapBatches(Fn)]. These stages cannot
+    be fused, since read stages are handled specially.
+    After rewriting to [GetReadTasks -> MapBatches(DoRead) -> MapBatches(Fn)],
+    now we can fuse the latter two MapBatches stages into a single OneToOne stage:
+    [GetReadTasks -> MapBatches(DoRead -> Fn)].
+
+    Args:
+        blocks: Lazy block list representing read stage.
+
+    Returns:
+        Non-lazy block list containing read tasks for not-yet-read block partitions,
+        new stats for the block list, and the new one-to-one read stage.
+    """
+    # Generate the "GetReadTasks" stage blocks.
+    remote_args = in_blocks._remote_args
+    blocks, metadata = [], []
+    for read_task in in_blocks._tasks:
+        blocks.append(ray.put(read_task._read_fn))
+        metadata.append(read_task.get_metadata())
+    block_list = BlockList(blocks, metadata)
+
+    def block_fn(read_fn: Callable[[], Iterator[Block]]) -> Iterator[Block]:
+        for block in read_fn():
+            yield block
+
+    stage = OneToOneStage("read", block_fn, "tasks", remote_args)
+    stats = DatasetStats(stages={}, parent=None)
+    return block_list, stats, stage
+
+
+def _fuse_one_to_one_stages(stages: List[Stage]) -> List[Stage]:
+    """Fuses compatible one-to-one stages.
+
+    Args:
+        stages: Stages to try to fuse.
+
+    Returns:
+        Optimized, fused stages.
+    """
+    optimized_stages = []
```

Comment:
Just 'fused_stages'? There is no optimization other than fusing compatible stages.

Reply:
I suppose that this function is only concerned with fusing stages, even if we might add future optimizations. I'll make that change!